File: src/lib/server/seed.ts
/**
 * Демо-данные дежурств и ночных вожатых.
 *
 * График дежурств и ночные вожатые онлайн НЕ публикуются (разделы 3.4–3.5
 * документа) — в реальной эксплуатации их вносит администратор через
 * POST /api/duty и POST /api/counselors. Здесь — примерные данные,
 * чтобы интерфейс не выглядел пустым при первом запуске.
 */

import { db } from "@/lib/db";

const DUTY_ROTATION: Array<{ className: string; interval: string; responsible: string }> = [
  { className: "10-1", interval: "после 2-й пары (полдник)", responsible: "Смирнова О. В." },
  { className: "10-2", interval: "после 3-й пары (обед)", responsible: "Ковалёв А. И." },
  { className: "10-3", interval: "после 5-й пары (ужин)", responsible: "Смирнова О. В." },
  { className: "9-1", interval: "после 2-й пары (полдник)", responsible: "Титова Н. С." },
  { className: "9-2", interval: "после 3-й пары (обед)", responsible: "Ковалёв А. И." },
  { className: "11-1", interval: "после 5-й пары (ужин)", responsible: "Гусев П. Е." },
  { className: "11-3", interval: "после 2-й пары (полдник)", responsible: "Титова Н. С." },
];

const NIGHT_ROTATION: Array<{ dormitory: string; name: string; floor: string }> = [
  { dormitory: "Общежитие №1", name: "Смирнова Ольга Викторовна", floor: "2–3 этажи" },
  { dormitory: "Общежитие №2", name: "Ковалёв Артём Игоревич", floor: "3–4 этажи" },
];

function fmt(date: Date): string {
  return `${String(date.getDate()).padStart(2, "0")}.${String(date.getMonth() + 1).padStart(2, "0")}.${date.getFullYear()}`;
}

let seedPromise: Promise<void> | null = null;

/** Однократный идемпотентный сид (при пустых таблицах) */
export async function ensureSeedData(): Promise<void> {
  if (!seedPromise) {
    seedPromise = (async () => {
      const dutyCount = await db.dutyEntry.count();
      const nightCount = await db.nightCounselor.count();
      if (dutyCount > 0 || nightCount > 0) return;

      const today = new Date();
      const dutyRows: Array<{
        date: string;
        dutyType: string;
        className: string;
        responsible: string;
        timeInterval: string;
        notes: string;
      }> = [];
      const nightRows: Array<{
        date: string;
        dormitory: string;
        counselorName: string;
        phone: string;
        floor: string;
        notes: string;
      }> = [];

      for (let offset = 0; offset < 14; offset++) {
        const day = new Date(today.getFullYear(), today.getMonth(), today.getDate() + offset);
        const weekday = day.getDay(); // 0=Вс

        // Ночные вожатые — ежедневно в обоих общежитиях (включая воскресенье)
        for (const night of NIGHT_ROTATION) {
          nightRows.push({
            date: fmt(day),
            dormitory: night.dormitory,
            counselorName: night.name,
            phone: "+7 (383) 373-96-41",
            floor: night.floor,
            notes: "Круглосуточное дежурство с обходами (демо-запись)",
          });
        }

        // Дежурства классов — только в учебные дни
        if (weekday === 0) continue;

        const rotation = DUTY_ROTATION[(weekday + offset) % DUTY_ROTATION.length];
        dutyRows.push({
          date: fmt(day),
          dutyType: "столовая",
          className: rotation.className,
          responsible: rotation.responsible,
          timeInterval: rotation.interval,
          notes: "Демо-запись: замените реальными данными через панель администратора",
        });

        if (weekday === 3) {
          dutyRows.push({
            date: fmt(day),
            dutyType: "корпус",
            className: "10-5",
            responsible: "Гусев П. Е.",
            timeInterval: "вечер, после 5-й пары",
            notes: "Дежурство по учебному корпусу (демо-запись)",
          });
        }
      }

      if (dutyRows.length) await db.dutyEntry.createMany({ data: dutyRows });
      if (nightRows.length) await db.nightCounselor.createMany({ data: nightRows });
    })().catch((error) => {
      seedPromise = null;
      throw error;
    });
  }
  return seedPromise;
}

File: fastapi-backend/seed.py
"""Наполнение SQLite демо-данными: дежурства классов и ночные вожатые.

График дежурств и ночные вожатые онлайн НЕ публикуются (см. анализ,
разделы 3.4–3.5), поэтому в реальной эксплуатации эти таблицы наполняет
администратор через ``POST /api/duty`` и ``POST /api/night-counselors``.
Этот скрипт создаёт ПРИМЕРНЫЕ данные на 7 дней вперёд, чтобы интерфейс
не выглядел пустым при первом запуске.

Запуск::

    python seed.py           # добавить демо-данные (сегодня + 6 дней)
    python seed.py --clear   # удалить ВСЕ дежурства и вожатых
"""

from __future__ import annotations

import datetime as dt
import sys

import database


def _fmt(day: dt.date) -> str:
    return day.strftime("%d.%m.%Y")


# Классы, дежурящие по столовой (реальные классы СУНЦ НГУ 2026/27)
_DUTY_CLASSES: list[tuple[str, str, str]] = [
    # (класс, время, ответственный воспитатель)
    ("10-1", "после 2-й пары (полдник)", "Смирнова О. В."),
    ("10-2", "после 3-й пары (обед)", "Ковалёв А. И."),
    ("10-3", "после 5-й пары (ужин)", "Смирнова О. В."),
    ("9-1", "после 2-й пары (полдник)", "Титова Н. С."),
    ("9-2", "после 3-й пары (обед)", "Ковалёв А. И."),
    ("11-1", "после 5-й пары (ужин)", "Гусев П. Е."),
    ("11-3", "после 2-й пары (полдник)", "Титова Н. С."),
]

# Ночные вожатые по общежитиям (ДЕМО-ФИО; заменить реальными при вводе)
_NIGHT: list[tuple[str, str, str, str]] = [
    # (общежитие, ФИО, телефон, этаж)
    ("Общежитие №1", "Смирнова Ольга Викторовна", "+7 (383) 373-96-41", "2–3 этажи"),
    ("Общежитие №2", "Ковалёв Артём Игоревич", "+7 (383) 373-96-41", "3–4 этажи"),
]


def seed(days: int = 7) -> tuple[int, int]:
    """Заполнить таблицы демо-данными на ``days`` дней вперёд. Возвращает (дежурства, вожатые)."""
    database.init_db()
    today = dt.date.today()
    duty_count = 0
    night_count = 0

    for offset in range(days):
        day = today + dt.timedelta(days=offset)
        weekday = day.weekday()  # 0 = Пн
        if weekday >= 6:  # воскресенье — дежурств нет
            continue

        # Дежурство класса по столовой — ротация по дням недели
        cls, interval, responsible = _DUTY_CLASSES[(weekday + offset) % len(_DUTY_CLASSES)]
        database.upsert_duty(
            {
                "date": _fmt(day),
                "duty_type": "столовая",
                "class_name": cls,
                "responsible": responsible,
                "time_interval": interval,
                "notes": "Помощь при раздаче, уборка столов (демо-запись)",
            }
        )
        duty_count += 1

        # Дежурство по корпусу — 10-е классы по средам
        if weekday == 2:
            database.upsert_duty(
                {
                    "date": _fmt(day),
                    "duty_type": "корпус",
                    "class_name": "10-5",
                    "responsible": "Гусев П. Е.",
                    "time_interval": "вечер, после 5-й пары",
                    "notes": "Дежурство по учебному корпусу (демо-запись)",
                }
            )
            duty_count += 1

        # Ночные вожатые — ежедневно в обоих общежитиях
        for dorm, name, phone, floor in _NIGHT:
            database.upsert_night_counselor(
                {
                    "date": _fmt(day),
                    "dormitory": dorm,
                    "counselor_name": name,
                    "phone": phone,
                    "floor": floor,
                    "notes": "Круглосуточное дежурство с обходами (демо-запись)",
                }
            )
            night_count += 1

    return duty_count, night_count


def clear() -> None:
    """Удалить все записи дежурств и ночных вожатых."""
    from sqlalchemy import delete

    database.init_db()
    with database.SessionLocal() as session:
        session.execute(delete(database.DutyEntryModel))
        session.execute(delete(database.NightCounselorModel))
        session.commit()


if __name__ == "__main__":
    if "--clear" in sys.argv:
        clear()
        print("Таблицы дежурств и ночных вожатых очищены.")
    else:
        duties, nights = seed()
        print(f"Добавлено демо-записей: дежурств — {duties}, ночных вожатых — {nights}.")
        print("Не забудьте заменить их реальными данными через POST /api/duty и")
        print("POST /api/night-counselors (заголовок X-Admin-Key).")
